cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java

```diff
@@ -0,0 +1,310 @@
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.spi.data.InvalidFieldException;
+import co.cask.cdap.spi.data.StructuredRow;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.table.StructuredTableSchema;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.FieldFactory;
+import co.cask.cdap.spi.data.table.field.FieldType;
+import co.cask.cdap.spi.data.table.field.Range;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Nosql structured table implementation.
+ */
+public final class NoSqlStructuredTable implements StructuredTable {
+  private static final Logger LOG = LoggerFactory.getLogger(NoSqlStructuredTable.class);
+  private final Table table;
+  private final StructuredTableSchema schema;
+  private final FieldFactory fieldFactory;
+
+  public NoSqlStructuredTable(Table table, StructuredTableSchema schema) {
+    this.table = table;
+    this.schema = schema;
+    this.fieldFactory = new FieldFactory(schema);
+  }
+
+  @Override
+  public void upsert(Collection<Field<?>> fields) throws InvalidFieldException {
+    LOG.trace("Table {}: Write fields {}", schema.getTableId(), fields);
+    Put put = convertFieldsToBytes(fields);
+    table.put(put.key, put.columns, put.values);
+  }
+
+  @Override
+  public Optional<StructuredRow> read(Collection<Field<?>> keys) throws InvalidFieldException {
+    return read(keys, Collections.emptySet());
+  }
+
+  @Override
+  public Optional<StructuredRow> read(Collection<Field<?>> keys,
+                                      Collection<String> columns) throws InvalidFieldException {
+    LOG.trace("Table {}: Read with keys {} and columns", schema.getTableId(), keys, columns);
+    co.cask.cdap.api.dataset.table.Row row = table.get(convertKeyToBytes(keys, false),
+                                                       convertColumnsToBytes(columns));
+    return row.isEmpty() ? Optional.empty() : Optional.of(new NoSqlStructuredRow(row, schema));
+  }
+
+  @Override
+  public CloseableIterator<StructuredRow> scan(Range keyRange, int limit) throws InvalidFieldException {
+    LOG.trace("Table {}: Scan range {} with limit {}", schema.getTableId(), keyRange, limit);
+    byte[] begin = keyRange.getBegin() == null ? null : convertKeyToBytes(keyRange.getBegin(), true);
+    byte[] end = keyRange.getEnd() == null ? null : convertKeyToBytes(keyRange.getEnd(), true);
+
+    // Table.scan() start key is inclusive by default
+    if (begin != null && keyRange.getBeginBound() == Range.Bound.EXCLUSIVE) {
+      begin = Bytes.stopKeyForPrefix(begin);
```

Comment:
I don't think this is correct. Consider a scan with a prefix partial key with "k1" as the begin field. By calculating the `begin` with the `stopKeyForPrefix`, it will effective skip all the rows that has "k1" as prefix.

Reply:
Yea but here the condition is EXCLUSIVE for the begin prefix, so it should skip these prefixes